src/org/exist/http/urlrewrite/XQueryURLRewrite.java

```diff
@@ -726,166 +728,166 @@ protected String adjustPathForSourceLookup(String basePath, String path) {
-              return null;
             }
-        } else {
+
             if (LOG.isTraceEnabled()) {
-            	LOG.trace("Looking for controller.xql in the filesystem, starting from: " + basePath);
+                LOG.trace("Found controller file: " + controllerDoc.getURI());
             }
-            final String realPath = config.getServletContext().getRealPath(basePath);
-            final Path baseDir = Paths.get(realPath);
-            if (!Files.isDirectory(baseDir)) {
-                LOG.warn("Base path for XQueryURLRewrite does not point to a directory");
+
+            if (controllerDoc.getResourceType() != DocumentImpl.BINARY_FILE ||
+                    !"application/xquery".equals(controllerDoc.getMetadata().getMimeType())) {
+                LOG.warn("XQuery resource: " + query + " is not an XQuery or declares a wrong mime-type");
                 return null;
             }
 
-            Path controllerFile = null;
-            Path subDir = baseDir;
-            for (int i = 0; i < components.length; i++) {
-                if (components[i].length() > 0) {
-                    subDir = subDir.resolve(components[i]);
-                    if (Files.isDirectory(subDir)) {
-                        Path cf = subDir.resolve("controller.xql");
-                        if (Files.isReadable(cf)) {
-                            controllerFile = cf;
-                        }
-                    } else {
-                        break;
-                    }
-                }
-            }
-            if (controllerFile == null) {
-                Path cf = baseDir.resolve("controller.xql");
-                if (Files.isReadable(cf)) {
-                    controllerFile = cf;
-                }
+            final String controllerPath = controllerDoc.getCollection().getURI().getRawCollectionPath();
+            final SourceInfo sourceInfo = new SourceInfo(new DBSource(broker, (BinaryDocument) controllerDoc, true), "xmldb:exist://" + controllerPath);
+            sourceInfo.controllerPath = controllerPath.substring(locationUri.getCollectionPath().length());
+            return sourceInfo;
+
+        } catch (final URISyntaxException e) {
+            LOG.warn("Bad URI for base path: " + e.getMessage(), e);
+            return null;
+        } finally {
+            if (controllerDoc != null) {
+                controllerDoc.getUpdateLock().release(LockMode.READ_LOCK);
             }
-            if (controllerFile == null) {
-                LOG.warn("XQueryURLRewrite controller could not be found");
+        }
+    }
+
+    /**
+     * Finds a `controller.xql` file within a Collection hierarchy.
+     * Most specific collections are considered first.
+     *
+     * For example, given the collectionUri `/db/apps`
+     * and the pathPomponents `['myapp', 'data']`, the
+     * order or search will be:
+     *
+     * /db/apps/myapp/data/collection.xconf
```

Comment:
@adamretter Am I correct that in the original code the search is actually reverse?
```
/db/apps/collection.xconf
/db/apps/myapp/collection.xconf
/db/apps/myapp/data/collection.xconf
``` 

Your approach makes more sense to me in a way, but I am not sure bout the consequences? @wolfgangmm

Reply:
Nope. They both search in the same direction. As I mentioned, the original code is very hard to understand

Reply:
I confirm.... the code was unbelievable ugly.